[Bug]: Master failed to apply snapshot when `/tmp` and master data dir in different block dervice

### Contact Details

huangzhibin1@oppo.com

### Is there an existing issue for this?

- [X] I have searched all the existing issues

### Priority

low (Default)

### Environment

```markdown
- CubeFS version:
- Deployment mode(docker or standalone or cluster):
- Dependent components:
- OS kernel version(Ubuntu or CentOS):
- CPU/Memory:
- Others:
```


### Current Behavior

2024/02/28 11:06:02.146748 [ERROR] metadata_fsm.go:280: action[ApplySnapshot] failed,err:rename /home/service/var/data/cfs/master/store /tmp/remove_by_rename2989293880/remove: invalid cross-device link

### Expected Behavior

Apply snapshot successfully

### Steps To Reproduce

_No response_

### CubeFS Log

_No response_

### Anything else? (Additional Context)

_No response_

--- a/master/metadata_fsm.go
+++ b/master/metadata_fsm.go
@@ -19,6 +19,7 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path"
 	"strconv"
 
 	"github.com/cubefs/cubefs/depends/tiglabs/raft"
@@ -206,6 +207,7 @@ func (mf *MetadataFsm) ApplySnapshot(peers []proto.Peer, iterator proto.SnapIter
 	}
 	rocksdbOpened := true
 	removeDir := ""
+	garbageDir := ""
 	// open temp rocksdb
 	tempDb, err := raftstore.NewRocksDBStore(recoveryDir, mf.store.GetLruCacheSize(), mf.store.GetWriteBufferSize())
 	if err != nil {
@@ -216,12 +218,19 @@ func (mf *MetadataFsm) ApplySnapshot(peers []proto.Peer, iterator proto.SnapIter
 	mf.store.Close()
 	rocksdbOpened = false
 	// remove by rename
-	removeDir, err = os.MkdirTemp("", "remove_by_rename")
-	if err != nil {
-		log.LogErrorf("failed to get temp dir %v", err.Error())
+	removeDir = raftstore.GetRocksDBStoreGarbageDir(mf.store.GetDir())
+	err = os.MkdirAll(removeDir, 0o755)
+	if err != nil && !os.IsExist(err) {
+		log.LogErrorf("[ApplySnapshot] failed to create garbage dir, err(%v)", err)
 		goto errHandler
 	}
-	removeDir = fmt.Sprintf("%s/remove", removeDir)
+	removeDir, err = os.MkdirTemp(removeDir, "")
+	if err != nil {
+		log.LogErrorf("[ApplySnapshot] failed to get temp dir %v", err.Error())
+		goto errHandler
+	}
+	garbageDir = removeDir
+	removeDir = path.Join(removeDir, "remove_by_rename")
 	log.LogWarnf(fmt.Sprintf("action[ApplySnapshot] begin,applied[%v]", mf.applied))
 	for err == nil {
 		bgTime := stat.BeginStat()
@@ -256,9 +265,9 @@ func (mf *MetadataFsm) ApplySnapshot(peers []proto.Peer, iterator proto.SnapIter
 	if err = os.Rename(mf.store.GetDir(), removeDir); err != nil {
 		goto errHandler
 	}
-	if err = os.RemoveAll(removeDir); err != nil {
+	if err = os.RemoveAll(garbageDir); err != nil {
 		err = nil
-		log.LogErrorf("failed to remove directory %v", err.Error())
+		log.LogErrorf("[ApplySnapshot] failed to remove directory(%v) please try to remove it manually, err(%v)", garbageDir, err.Error())
 	}
 	// rename new dir to raft store dir
 	if err = os.Rename(tempDb.GetDir(), mf.store.GetDir()); err != nil {

--- a/raftstore/raftstore_db/store_rocksdb.go
+++ b/raftstore/raftstore_db/store_rocksdb.go
@@ -66,6 +66,11 @@ func GetRocksDBStoreRecoveryDir(dir string) string {
 	return fmt.Sprintf("%v_temp", dir)
 }
 
+func GetRocksDBStoreGarbageDir(dir string) string {
+	dir = strings.TrimSuffix(dir, "/")
+	return fmt.Sprintf("%v_garbage", dir)
+}
+
 // NewRocksDBStoreAndRecovery returns a new RocksDB instance after execute recovery.
 func NewRocksDBStoreAndRecovery(dir string, lruCacheSize, writeBufferSize int) (store *RocksDBStore, err error) {
 	// start recovery

